src/oss/langgraph/workflows-agents.mdx

```diff
@@ -0,0 +1,2229 @@
+output = structured_llm.invoke("How does Calcium CT score relate to high cholesterol?")
+
+# Define a tool
+def multiply(a: int, b: int) -> int:
+    return a * b
+
+# Augment the LLM with tools
+llm_with_tools = llm.bind_tools([multiply])
+
+# Invoke the LLM with input that triggers the tool call
+msg = llm_with_tools.invoke("What is 2 times 3?")
+
+# Get the tool call
+msg.tool_calls
+```
+:::
+:::js
+```typescript
+
+import { z } from "zod";
+import { tool } from "langchain";
+
+// Schema for structured output
+const SearchQuery = z.object({
+  search_query: z.string().describe("Query that is optimized web search."),
+  justification: z
+    .string()
+    .describe("Why this query is relevant to the user's request."),
+});
+
+// Augment the LLM with schema for structured output
+const structuredLlm = llm.withStructuredOutput(SearchQuery);
+
+// Invoke the augmented LLM
+const output = await structuredLlm.invoke(
+  "How does Calcium CT score relate to high cholesterol?"
+);
+
+// Define a tool
+const multiply = tool(
+  ({ a, b }) => {
+    return a * b;
+  },
+  {
+    name: "multiply",
+    description: "Multiply two numbers",
+    schema: z.object({
+      a: z.number(),
+      b: z.number(),
+    }),
+  }
+);
+
+// Augment the LLM with tools
+const llmWithTools = llm.bindTools([multiply]);
+
+// Invoke the LLM with input that triggers the tool call
+const msg = await llmWithTools.invoke("What is 2 times 3?");
+
+// Get the tool call
+console.log(msg.tool_calls);
+```
+:::
+
+## Prompt chaining
+
+Prompt chaining is when each LLM call processes the output of the previous call. It's often used for performing well-defined tasks that can be broken down into smaller, verifiable steps. Some examples include translating documents into different languages or verifying generated content for consistency.
+
+![Prompt chaining](/oss/images/prompt_chain.png)
+
+### Code examples
+
+:::python
+<Tabs>
+<Tab title="Graph API">
+```python
+from typing_extensions import TypedDict
+from langgraph.graph import StateGraph, START, END
+from IPython.display import Image, display
+
+
+# Graph state
+class State(TypedDict):
+    topic: str
+    joke: str
+    improved_joke: str
+    final_joke: str
+
+
+# Nodes
+def generate_joke(state: State):
+    """First LLM call to generate initial joke"""
+
+    msg = llm.invoke(f"Write a short joke about {state['topic']}")
+    return {"joke": msg.content}
+
+
+def check_punchline(state: State):
+    """Gate function to check if the joke has a punchline"""
+
+    # Simple check - does the joke contain "?" or "!"
+    if "?" in state["joke"] or "!" in state["joke"]:
+        return "Pass"
+    return "Fail"
+
+
+def improve_joke(state: State):
+    """Second LLM call to improve the joke"""
+
+    msg = llm.invoke(f"Make this joke funnier by adding wordplay: {state['joke']}")
+    return {"improved_joke": msg.content}
+
+
+def polish_joke(state: State):
+    """Third LLM call for final polish"""
+
```

Comment:
This extra new line after a doc-string is not consistent with styling of code that we use in any our code bases.

```suggestion
```

I'd suggest removing it if easy to do quickly (or at least not to introduce it into further docs)

(Ideally, we'll be able to auto format this soon with ruff -- just need to set up a bit of tooling)